src/policies/cooler/CoolerV2Migrator.sol

```diff
@@ -0,0 +1,439 @@
+        FLASH.flashLoan(
+            this,
+            address(DAI),
+            daiRequired + usdsRequired,
+            abi.encode(
+                FlashLoanData(coolerData, msg.sender, newOwner_, usdsRequired, delegationRequests_)
+            )
+        );
+
+        // This shouldn't happen, but transfer any leftover funds back to the sender
+        uint256 usdsBalanceAfter = USDS.balanceOf(address(this));
+        if (usdsBalanceAfter > 0) {
+            USDS.safeTransfer(msg.sender, usdsBalanceAfter);
+        }
+        uint256 daiBalanceAfter = DAI.balanceOf(address(this));
+        if (daiBalanceAfter > 0) {
+            DAI.safeTransfer(msg.sender, daiBalanceAfter);
+        }
+    }
+
+    /// @inheritdoc IERC3156FlashBorrower
+    /// @dev        This function reverts if:
+    ///             - The caller is not the flash loan provider
+    ///             - The initiator is not this contract
+    function onFlashLoan(
+        address initiator_,
+        address, // flashloan token is only DAI
+        uint256 amount_,
+        uint256 lenderFee_,
+        bytes calldata params_
+    ) external override returns (bytes32) {
+        // perform sanity checks
+        if (msg.sender != address(FLASH)) revert OnlyLender();
+        if (initiator_ != address(this)) revert OnlyThis();
+
+        // Unpack param data
+        FlashLoanData memory flashLoanData = abi.decode(params_, (FlashLoanData));
+        CoolerData[] memory coolers = flashLoanData.coolers;
+
+        // If there are loans in USDS, convert the required amount to DAI
+        if (flashLoanData.usdsRequired > 0) {
+            DAI.safeApprove(address(MIGRATOR), flashLoanData.usdsRequired);
+            MIGRATOR.daiToUsds(address(this), flashLoanData.usdsRequired);
+        }
+
+        // Keep track of debt tokens out and collateral in
+        uint256 totalRepaid;
+        uint256 totalCollateral;
+
+        // Validate and repay loans for each cooler
+        for (uint256 i; i < coolers.length; ++i) {
+            CoolerData memory coolerData = coolers[i];
+
+            (uint256 repaid, uint256 collateral) = _handleRepayments(
+                coolerData.cooler,
+                coolerData.debtToken,
+                coolerData.numLoans
+            );
+            totalRepaid += repaid;
+            totalCollateral += collateral;
+        }
+
+        // Transfer the collateral from the cooler owner to this contract
+        GOHM.safeTransferFrom(flashLoanData.currentOwner, address(this), totalCollateral);
+
+        // Approve the Cooler V2 to spend the collateral
+        GOHM.safeApprove(address(COOLERV2), totalCollateral);
+
+        // Add collateral and borrow spent flash loan from Cooler V2
+        COOLERV2.addCollateral(
+            totalCollateral.encodeUInt128(),
+            flashLoanData.newOwner,
+            flashLoanData.delegationRequests
+        );
+        COOLERV2.borrow(
+            (totalRepaid + lenderFee_).encodeUInt128(),
```

Comment:
Has analysis been done already that all cooler loans would be under the starting OLTV that we'd set on v2 rollout?

If there was a flash fee, this could bring them over the OLTV and this would revert?

Reply:
Fair point. I had _assumed_ that the OLTV would be higher, in line with the increased backing. The existing Clearinghouses have had it set at the same value - `2,892.92` - whereas current backing/gOHM is `3,141.28`.

If that isn't the case: then both interest and lender fee should be paid by the caller, since the interest could tip the borrow amount over the limit.

Reply:
Do you think there should be a provision (e.g. boolean) for the user paying the interest + lender fee and avoiding borrowing anything additional from the MonoCooler?

Reply:
Added this. Caller can opt to pay, in which case they provide the amount in DAI (easier as the flash loan is in DAI). `previewConsolidate()` updated as well.